librz/util/print.c

```diff
@@ -1111,62 +1184,87 @@ static ut8 *M(const ut8 *b, int len) {
+ * \return Hexdump diff string.
+ */
 // TODO: add support for cursor
-RZ_API void rz_print_hexdiff(RzPrint *p, ut64 aa, const ut8 *_a, ut64 ba, const ut8 *_b, int len, int scndcol) {
+RZ_API RZ_OWN char *rz_print_hexdiff_str(RZ_NONNULL RzPrint *p, ut64 aa, RZ_NONNULL const ut8 *_a, ut64 ba, RZ_NONNULL const ut8 *_b, int len, int scndcol) {
+	rz_return_val_if_fail(p && _a && _b && len > 0, NULL);
 	ut8 *a, *b;
 	char linediff, fmt[64];
 	int color = p->flags & RZ_PRINT_FLAGS_COLOR;
 	int diffskip = p->flags & RZ_PRINT_FLAGS_DIFFOUT;
 	int i, j, min;
+	RzStrBuf *sb = rz_strbuf_new(NULL);
 	if (!((a = M(_a, len)))) {
-		return;
+		return NULL;
 	}
 	if (!((b = M(_b, len)))) {
 		free(a);
-		return;
+		return NULL;
 	}
 	for (i = 0; i < len; i += 16) {
 		min = RZ_MIN(16, len - i);
 		linediff = (memcmp(a + i, b + i, min)) ? '!' : '|';
 		if (diffskip && linediff == '|') {
 			continue;
 		}
-		p->cb_printf("0x%08" PFMT64x " ", aa + i);
+		rz_strbuf_appendf(sb, "0x%08" PFMT64x " ", aa + i);
 		for (j = 0; j < min; j++) {
 			*fmt = color;
-			rz_print_cursor(p, i + j, 1, 1);
-			p->cb_printf("%s", BD(a, b));
-			rz_print_cursor(p, i + j, 1, 0);
+			rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 1));
+			rz_strbuf_appendf(sb, "%s", BD(a, b));
+			rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 0));
 		}
-		p->cb_printf(" ");
+		rz_strbuf_append(sb, " ");
 		for (j = 0; j < min; j++) {
 			*fmt = color;
-			rz_print_cursor(p, i + j, 1, 1);
-			p->cb_printf("%s", CD(a, b));
-			rz_print_cursor(p, i + j, 1, 0);
+			rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 1));
+			rz_strbuf_appendf(sb, "%s", CD(a, b));
+			rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 0));
 		}
 		if (scndcol) {
-			p->cb_printf(" %c 0x%08" PFMT64x " ", linediff, ba + i);
+			rz_strbuf_appendf(sb, " %c 0x%08" PFMT64x " ", linediff, ba + i);
 			for (j = 0; j < min; j++) {
 				*fmt = color;
-				rz_print_cursor(p, i + j, 1, 1);
-				p->cb_printf("%s", BD(b, a));
-				rz_print_cursor(p, i + j, 1, 0);
+				rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 1));
+				rz_strbuf_appendf(sb, "%s", BD(b, a));
+				rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 0));
 			}
-			p->cb_printf(" ");
+			rz_strbuf_append(sb, " ");
 			for (j = 0; j < min; j++) {
 				*fmt = color;
-				rz_print_cursor(p, i + j, 1, 1);
-				p->cb_printf("%s", CD(b, a));
-				rz_print_cursor(p, i + j, 1, 0);
+				rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 1));
+				rz_strbuf_appendf(sb, "%s", CD(b, a));
+				rz_strbuf_append(sb, rz_print_cursor_str(p, i + j, 1, 0));
 			}
-			p->cb_printf("\n");
+			rz_strbuf_append(sb, "\n");
 		} else {
-			p->cb_printf(" %c\n", linediff);
+			rz_strbuf_appendf(sb, " %c\n", linediff);
 		}
 	}
 	free(a);
 	free(b);
+	return rz_strbuf_drain(sb);
+}
+
+/// \see rz_print_hexdiff_str
+RZ_API void rz_print_hexdiff(RZ_NONNULL RzPrint *p, ut64 aa, RZ_NONNULL const ut8 *_a, ut64 ba, RZ_NONNULL const ut8 *_b, int len, int scndcol) {
```

Comment:
I suggest removing this function from `librz/util/` and use `rz_print_hexdiff_str()` wherever appropriate directly.